go/adbc/driver/snowflake/bulk_ingestion.go

```diff
@@ -0,0 +1,546 @@
+	readyFn := func() {
+
+		// readyFn is a no-op if the shutdown signal has already been recieved
+		select {
+		case _, ok := <-stopCh:
+			if !ok {
+				return
+			}
+		default:
+			// Continue
+		}
+
+		// readyFn is a no-op if it already knows that at least 1 file is ready to be loaded
+		select {
+		case readyCh <- struct{}{}:
+		default:
+			return
+		}
+	}
+
+	// Handler to signal that ingestion pipeline has completed successfully.
+	// Executes COPY to finalize ingestion (may no-op if all files already loaded by bg workers) and waits for all COPYs to complete.
+	stopFn := func() error {
+		defer cancel()
+		close(stopCh)
+		close(readyCh)
+
+		g.Go(func() error {
+			_, err := cn.ExecContext(ctx, copyQuery, []driver.NamedValue{{Value: tableName}})
+			return err
+		})
+
+		return g.Wait()
+	}
+
+	// Handler to signal that ingestion pipeline failed and COPY operations should not proceed.
+	// Stops the dispatch of new bg workers and cancels all in-progress COPY operations.
+	cancelFn := func() {
+		defer cancel()
+		close(stopCh)
+		close(readyCh)
+	}
+
+	go func() {
+		for {
+
+			// Check for shutdown signals before starting a new COPY
+			select {
+			case <-stopCh:
+				return
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			g.Go(func() error {
+				// Block until there is at least 1 new file available for copy, or it's time to shutdown
+				select {
+				case _, ok := <-readyCh:
+					if !ok {
+						return nil
+					}
+					// Proceed to start a new COPY job
+				case <-ctx.Done():
+					return ctx.Err()
+				}
+
+				_, err := cn.ExecContext(ctx, copyQuery, []driver.NamedValue{{Value: tableName}})
+
+				return err
+			})
+		}
+	}()
+
+	return readyFn, stopFn, cancelFn
+}
+
+func countRowsInTable(ctx context.Context, db *sql.DB, tableName string) (int64, error) {
+	var nrows int64
+
+	row := db.QueryRowContext(ctx, countQuery, tableName)
+	if err := row.Scan(&nrows); err != nil {
+		return 0, err
+	}
+
+	return nrows, nil
+}
+
+// Initializes a sync.Pool of *bytes.Buffer.
+// Extra space is preallocated so that the Parquet footer can be written after reaching target file size without growing the buffer
+func newBufferPool(size int) (getFn func() *bytes.Buffer, putFn func(*bytes.Buffer)) {
+	buffers := sync.Pool{
+		New: func() interface{} {
+			extraSpace := 1 * megabyte // TODO(joellubi): Generally works, but can this be smarter?
+			buf := make([]byte, 0, size+extraSpace)
+			return bytes.NewBuffer(buf)
+		},
+	}
+
+	getFn = func() *bytes.Buffer {
+		return buffers.Get().(*bytes.Buffer)
+	}
+
+	putFn = func(buf *bytes.Buffer) {
+		buf.Reset()
+		buffers.Put(buf)
+	}
+
+	return getFn, putFn
+}
+
+// Wraps an io.Writer and specifies a limit.
+// Keeps track of how many bytes have been written and can report whether the limit has been exceeded.
+type limitWriter struct {
+	w     io.Writer
+	limit int
+
+	bytesWritten int
+}
```

Comment:
rather than tracking the bytes written like this, the parquet writer already exposes the ability to check how many bytes have been written via a `TotalBytesWritten` method. Having this just seems redundant.

Reply:
Ah thanks for pointing that out, I wasn't aware that bytes were being tracked by the writer. I did give this a try but it appears that the most similar method exposed by the parquet writer is `RowGroupTotalBytesWritten`. This is pretty close but unfortunately the count resets with each new row group. Since multiple row groups may be created in a single call to `Write` in the case of records larger than the max row group size, it doesn't seem there's a straightforward way to account for all bytes written with this approach.

Reply:
For situations like this, a single row group per file makes the most sense. We should set the max row group length to `math.MaxInt64`, at which point you can use the `RowGroupTotalBytesWritten` to handle it.

Reply:
This approach makes sense to me. I did try to add the parquet writer option `parquet.WithMaxRowGroupLength(math.MaxInt64)` but the value `RowGroupTotalBytesWritten` returns still doesn't seem to agree exactly with the total number of bytes the writer actually has going through it. It doesn't report as many bytes written for a given write as the limitWriter does. As a result the file sizes written using this approach are typically between 13-14 MB, whereas they were previously between 10-11 which is what I would expect. Any idea why this might be the case?

Reply:
I opened [arrow@39789](https://github.com/apache/arrow/issues/39789) to track this. I spent a bit of time looking into this but I think I'd like to isolate it with a test separately to help figure it out. Any objection to leaving the current approach with a reference to the issue, and following up with a fix separately?